overlord/devicestate/devicestate_systems_test.go

```diff
@@ -631,3 +631,159 @@ func (s *deviceMgrSystemsSuite) TestRequestAction1618(c *C) {
+			BrandID: s.mockedSystemSeeds[0].brand.AccountID(),
+		},
+	})
+	s.state.Unlock()
+
+	err := s.mgr.Reboot("20191119", "install")
+	c.Assert(err, IsNil)
+
+	m, err := s.bootloader.GetBootVars("snapd_recovery_mode", "snapd_recovery_system")
+	c.Assert(err, IsNil)
+	c.Check(m, DeepEquals, map[string]string{
+		"snapd_recovery_system": "20191119",
+		"snapd_recovery_mode":   "install",
+	})
+	c.Check(s.restartRequests, DeepEquals, []state.RestartType{state.RestartSystemNow})
+}
+
+func (s *deviceMgrSystemsSuite) TestRebootModeOnlyHappy(c *C) {
+	s.state.Lock()
+	s.state.Set("seeded-systems", []devicestate.SeededSystem{
+		{
+			System:  s.mockedSystemSeeds[0].label,
+			Model:   s.mockedSystemSeeds[0].model.Model(),
+			BrandID: s.mockedSystemSeeds[0].brand.AccountID(),
+		},
+	})
+	s.state.Unlock()
+
+	for _, mode := range []string{"recover", "install"} {
+		s.restartRequests = nil
+		s.bootloader.BootVars = make(map[string]string)
+
+		err := s.mgr.Reboot("", mode)
+		c.Assert(err, IsNil)
+
+		m, err := s.bootloader.GetBootVars("snapd_recovery_mode", "snapd_recovery_system")
+		c.Assert(err, IsNil)
+		c.Check(m, DeepEquals, map[string]string{
+			"snapd_recovery_system": s.mockedSystemSeeds[0].label,
+			"snapd_recovery_mode":   mode,
+		})
+		c.Check(s.restartRequests, DeepEquals, []state.RestartType{state.RestartSystemNow})
+	}
+}
+
+func (s *deviceMgrSystemsSuite) TestRebootFromRecoverToRun(c *C) {
+	modeenv := boot.Modeenv{
+		Mode:           "recover",
+		RecoverySystem: s.mockedSystemSeeds[0].label,
+	}
+	err := modeenv.WriteTo("")
+	c.Assert(err, IsNil)
+
+	devicestate.SetSystemMode(s.mgr, "recover")
+	err = s.bootloader.SetBootVars(map[string]string{
+		"snapd_recovery_mode":   "recover",
+		"snapd_recovery_system": s.mockedSystemSeeds[0].label,
+	})
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	s.state.Set("seeded-systems", []devicestate.SeededSystem{
+		{
+			System:  s.mockedSystemSeeds[0].label,
+			Model:   s.mockedSystemSeeds[0].model.Model(),
+			BrandID: s.mockedSystemSeeds[0].brand.AccountID(),
+		},
+	})
+	s.state.Unlock()
+
+	err = s.mgr.Reboot("", "run")
+	c.Assert(err, IsNil)
+
+	m, err := s.bootloader.GetBootVars("snapd_recovery_mode", "snapd_recovery_system")
+	c.Assert(err, IsNil)
+	c.Check(m, DeepEquals, map[string]string{
+		"snapd_recovery_mode":   "run",
+		"snapd_recovery_system": s.mockedSystemSeeds[0].label,
+	})
+	c.Check(s.restartRequests, DeepEquals, []state.RestartType{state.RestartSystemNow})
+}
+
+func (s *deviceMgrSystemsSuite) TestRebootAlreadyInRunMode(c *C) {
+	devicestate.SetSystemMode(s.mgr, "run")
+
+	s.state.Lock()
+	s.state.Set("seeded-systems", []devicestate.SeededSystem{
+		{
+			System:  s.mockedSystemSeeds[0].label,
+			Model:   s.mockedSystemSeeds[0].model.Model(),
+			BrandID: s.mockedSystemSeeds[0].brand.AccountID(),
+		},
+	})
+	s.state.Unlock()
+
+	// XXX: should this error instead with something like
+	//      ErrAlreadyInRequestedMode
+	err := s.mgr.Reboot("", "run")
```

Comment:
hmm, I don't think that should error but this behavior doesn't look correct either, I would expect this to behave the same as Reboot("", ""). Basically reboot should always reboot, no?

Reply:
This is fixed now, it behaves the same way as Reboot("","") now.